src/tools/rbd_mirror/InstanceReplayer.cc

```diff
@@ -541,6 +645,223 @@ void InstanceReplayer<I>::schedule_image_state_check_task() {
+      ++error_count;
+    }
+
+    start_group_replayer(current_it->second);
+  }
+
+  m_service_daemon->add_or_update_namespace_attribute(
+    m_local_io_ctx.get_id(), m_local_io_ctx.get_namespace(),
+    SERVICE_DAEMON_GROUP_ASSIGNED_COUNT_KEY, group_count);
+  m_service_daemon->add_or_update_namespace_attribute(
+    m_local_io_ctx.get_id(), m_local_io_ctx.get_namespace(),
+    SERVICE_DAEMON_GROUP_WARNING_COUNT_KEY, warning_count);
+  m_service_daemon->add_or_update_namespace_attribute(
+    m_local_io_ctx.get_id(), m_local_io_ctx.get_namespace(),
+    SERVICE_DAEMON_GROUP_ERROR_COUNT_KEY, error_count);
+
+  m_async_op_tracker.finish_op();
+}
+
+template <typename I>
+void InstanceReplayer<I>::stop_group_replayer(GroupReplayer<I> *group_replayer,
+                                              Context *on_finish) {
+  dout(10) << group_replayer << " global_group_id="
+           << group_replayer->get_global_group_id() << ", on_finish="
+           << on_finish << dendl;
+
+  if (group_replayer->is_stopped()) {
+    m_threads->work_queue->queue(on_finish, 0);
+    return;
+  }
+
+  m_async_op_tracker.start_op();
+  Context *ctx = create_async_context_callback(
+    m_threads->work_queue, new LambdaContext(
+      [this, group_replayer, on_finish] (int r) {
+        stop_group_replayer(group_replayer, on_finish);
+        m_async_op_tracker.finish_op();
+      }));
+
+  if (group_replayer->is_running()) {
+    group_replayer->stop(ctx, false);
+  } else {
+    int after = 1;
+    dout(10) << "scheduling group replayer " << group_replayer << " stop after "
+             << after << " sec (task " << ctx << ")" << dendl;
+    ctx = new LambdaContext(
+      [this, after, ctx] (int r) {
+        std::lock_guard timer_locker{m_threads->timer_lock};
+        m_threads->timer->add_event_after(after, ctx);
+      });
+    m_threads->work_queue->queue(ctx, 0);
+  }
+}
+
+template <typename I>
+void InstanceReplayer<I>::stop_group_replayers() {
+  dout(10) << dendl;
+
+  ceph_assert(ceph_mutex_is_locked(m_lock));
+
+  Context *ctx = create_async_context_callback(
+    m_threads->work_queue, create_context_callback<InstanceReplayer<I>,
+    &InstanceReplayer<I>::handle_stop_group_replayers>(this));
+
+  C_Gather *gather_ctx = new C_Gather(g_ceph_context, ctx);
+  for (auto &[_, group_replayer] : m_group_replayers) {
+    stop_group_replayer(group_replayer, gather_ctx->new_sub());
+  }
+  gather_ctx->activate();
+}
+
+template <typename I>
+void InstanceReplayer<I>::handle_stop_group_replayers(int r) {
+  dout(10) << "r=" << r << dendl;
+
+  ceph_assert(r == 0);
+
+  Context *on_finish = nullptr;
+  {
+    std::lock_guard locker{m_lock};
+
+    for (auto &[_, group_replayer] : m_group_replayers) {
+      ceph_assert(group_replayer->is_stopped());
+      group_replayer->destroy();
+    }
+    m_group_replayers.clear();
+
+    std::swap(on_finish, m_on_shut_down);
+  }
+  if (on_finish) {
+    on_finish->complete(r);
```

Comment:
This should only be called once - after both group and image replayers have been stopped.